### ISSUE TITLE
[Request]: Add PostgreSQL 17 Support to Hydra (Cloud Native Postgres Operator)

### ISSUE THREAD

**Issue by Shah-Aayush** [hydradatabase/hydra#272]
### What's wrong?

**Description:**  
PostgreSQL 17 introduces several key features like enhanced incremental backups and recovery, which are highly relevant for modern workflows. We request support for PostgreSQL 17 in Hydra to leverage these capabilities within Kubernetes/Openshift environments.  

**Request:**  
1. Update Hydra to support PostgreSQL 17.  
2. If possible, provide the original Hydra PostgreSQL image file so we can replicate steps in the Dockerfile and test PostgreSQL 17 integration ourselves.  
3. Suggest steps or guidance for creating a custom image with PostgreSQL 17.  
4. Share an approximate timeline for the official release of PostgreSQL 17 support in Hydra.  

Thank you for your attention! Let us know how we can help or contribute to testing pre-release builds.

**Comment by Shah-Aayush** [comment]
I’m unsure if this is the right place to request this feature, as I couldn’t create a discussion thread. Please let me know if there’s a specific procedure I should follow for this type of request.  
Thank you!

**Comment by Shah-Aayush** [comment]
I think pgxman is still not supporting postgres 17, so is there anyway to build the dockerfile without using pgxman ?

**Comment by ivan-v-kush** [comment]
The 1st issue is
```
checking for pg_config... /usr/bin/pg_config
configure: error: Citus is not compatible with the detected PostgreSQL version 17.
```

Then we get compilation errors.

As hydra is based on citus we can find there several MRs for support PG17, that can be used as a reference for Hydra PG17

https://github.com/citusdata/citus/pull/7699
https://github.com/citusdata/citus/pull/7700
https://github.com/citusdata/citus/issues/7708
etc.

**Comment by ivan-v-kush** [comment]
@wuputah @owenthereal @JerrySievert, other hydra maintainers do you need any help from us to support PG17?

**Comment by JerrySievert** [comment]
unfortunately, the existing tests do not run on pg17, and will need to be redone - see https://github.com/JerrySievert/hydra/commit/3614299e3b87ff2e59c1531d6f49a68bb000ee75 for all of the changes that were needed.

**Comment by ivan-v-kush** [comment]
thanks! I've got warning with your commit
```
columnar_customscan.c:2510:19: warning: incompatible pointer types passing 'PlanState *' (aka 'struct PlanState *') to parameter of type 'ExprContext *' (aka 'struct ExprContext *') [-Wincompatible-pointer-types]
  FreeExprContext(&node->ss.ps, false);
                  ^~~~~~~~~~~~
/usr/include/postgresql/17/server/executor/executor.h:541:42: note: passing argument to parameter 'econtext' here
extern void FreeExprContext(ExprContext *econtext, bool isCommit);
                                         ^

```

**Comment by ivan-v-kush** [comment]
@JerrySievert I suggest you create PR from your commit.

You need to use such construction to get `econtext`
```
FreeExprContext(css->ss.ps.ps_ExprContext, false);
```

**Comment by ivan-v-kush** [comment]
Is Docker Hydra's pipeline broken?

I try to launch tests using Docker, modified for this `docker-build.sh`. Added `--target checker` to launch checker stage.
```
docker build  --progress=plain --target checker -t $REPO:$TAG .
```

And got an error:
```
# initializing database system by copying initdb template
# copying of initdb template failed
# Examine "/columnar/src/test/regress/log/initdb.log" for the reason.
# Command was: cp -RPp "/tmp_install/initdb-template" "/columnar/src/test/regress/tmp_check/data" > "/columnar/src/test/regress/log/initdb.log" 2>&1
Bail out!make[1]: *** [Makefile:27: check-regression-columnar] Error 2
make[1]: Leaving directory '/columnar/src/test/regress'
make: *** [Makefile:49: check-all] Error 2
```

initdb.log contains
```
cp: cannot stat '/tmp_install/initdb-template': No such file or directory
```

**Comment by ivan-v-kush** [comment]
@JerrySievert do you mean this [by ](https://github.com/hydradatabase/hydra/issues/277) ?
> the existing tests do not run on pg17,

Or you ran them and `result` is different from `expected`?

**Comment by JerrySievert** [comment]
correct.  no PR's without tests to go with them.  test systems changed in pg16, which I adapted columnar's testing system to.  it changed again in pg17.

**Comment by ivan-v-kush** [comment]
The 2nd preparation PR
https://github.com/hydradatabase/hydra/pull/278

**Comment by ivan-v-kush** [comment]
Now in PG17 hydra has a halt problem when call `ANALYZE` sql command.
`columnar_scan_analyze_next_block` is called in analyze.
https://github.com/hydradatabase/hydra/blob/main/columnar/src/backend/columnar/columnar_tableam.c#L2066
It returns always `true`.
```
static bool
columnar_scan_analyze_next_block(TableScanDesc scan, BlockNumber blockno,
								 BufferAccessStrategy bstrategy)
{
	return true;
}
```
In PG17 code was refactored and now call of this function is in the loop (`table_scan_analyze_next_block` calls  `columnar_scan_analyze_next_block` for hydra)
https://github.com/postgres/postgres/blame/master/src/backend/commands/analyze.c#L1233
```
while (table_scan_analyze_next_block(scan, stream))
{
...
```

At the current time, I'm beginning to fix it. Citus seems also haven't fixed it.
https://github.com/citusdata/citus/blame/main/src/backend/columnar/columnar_tableam.c#L1427

**Comment by Vonng** [comment]
citus 13.0 already have PostgreSQL 17 support https://github.com/citusdata/citus/releases/tag/v13.0.1

Is there any update on bumping PG 17 support? 😊

**Comment by ivan-v-kush** [comment]
@Vonng it seems, they've moved OLAP to duckdb  from citus

https://www.hydra.so/blog-posts/duckdb-powered-postgres

**Comment by gregorywaynepower** [comment]
Here's a link that's still visible since the above link is broken: https://web.archive.org/web/20250119131606/https://www.hydra.so/blog-posts/duckdb-powered-postgres
